File: contrib/clojure-package/test/org/apache/clojure_mxnet/infer/objectdetector_test.clj
;; Licensed to the Apache Software Foundation (ASF) under one or more
;; contributor license agreements.  See the NOTICE file distributed with
;; this work for additional information regarding copyright ownership.
;; The ASF licenses this file to You under the Apache License, Version 2.0
;; (the "License"); you may not use this file except in compliance with
;; the License.  You may obtain a copy of the License at
;;
;;    http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.
;;

(ns org.apache.clojure-mxnet.infer.objectdetector-test
  (:require [org.apache.clojure-mxnet.context :as context]
            [org.apache.clojure-mxnet.dtype :as dtype]
            [org.apache.clojure-mxnet.infer :as infer]
            [org.apache.clojure-mxnet.layout :as layout]
            [clojure.java.io :as io]
            [clojure.java.shell :refer [sh]]
            [clojure.test :refer :all]
            [org.apache.clojure-mxnet.ndarray :as ndarray]))

(def model-dir "data/")
(def model-path-prefix (str model-dir "resnet50_ssd/resnet50_ssd_model"))

(when-not (.exists (io/file (str model-path-prefix "-symbol.json")))
  (sh "./scripts/infer/get_ssd_data.sh"))

(defn create-detector []
  (let [descriptors [{:name "data"
                      :shape [1 3 512 512]
                      :layout layout/NCHW
                      :dtype dtype/FLOAT32}]
        factory (infer/model-factory model-path-prefix descriptors)]
    (infer/create-object-detector factory)))

(deftest test-single-detection
  (let [detector (create-detector)
        image (infer/load-image-from-file "test/test-images/kitten.jpg")
        [predictions-all] (infer/detect-objects detector image)
        [predictions] (infer/detect-objects detector image 5)
        {:keys [class prob x-min x-max y-min y-max] :as pred} (first predictions)]
    (is (some? predictions))
    (is (= 5 (count predictions)))
    (is (= 13 (count predictions-all)))
    (is (= "cat" class))
    (is (< 0.8 prob))
    (every? #(< 0 % 1) [x-min x-max y-min y-max])))

(deftest test-batch-detection
  (let [detector (create-detector)
        image-batch (infer/load-image-paths ["test/test-images/kitten.jpg"
                                             "test/test-images/Pug-Cookie.jpg"])
        [batch-predictions-all] (infer/detect-objects-batch detector image-batch)
        [predictions] (infer/detect-objects-batch detector image-batch 5)
        {:keys [class prob x-min x-max y-min y-max] :as pred} (first predictions)]
    (is (some? predictions))
    (is (= 13 (count batch-predictions-all)))
    (is (= 5 (count predictions)))
    (is (= "cat" class))
    (is (< 0.8 prob))
    (every? #(< 0 % 1) [x-min x-max y-min y-max])))

(deftest test-detection-with-ndarrays
  (let [detector (create-detector)
        image (-> (infer/load-image-from-file "test/test-images/kitten.jpg")
                  (infer/reshape-image 512 512)
                  (infer/buffered-image-to-pixels [3 512 512] dtype/FLOAT32)
                  (ndarray/expand-dims 0))
        [predictions-all] (infer/detect-objects-with-ndarrays detector [image])
        [predictions] (infer/detect-objects-with-ndarrays detector [image] 1)
        {:keys [class prob x-min x-max y-min y-max] :as pred} (first predictions)]
        (is (some? predictions-all))
        (is (= 1 (count predictions)))
        (is (= "cat" class))
        (is (< 0.8 prob))
        (every? #(< 0 % 1) [x-min x-max y-min y-max])))

File: scala-package/infer/src/test/java/org/apache/mxnet/infer/javaapi/ObjectDetectorTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mxnet.infer.javaapi;

import org.apache.mxnet.Layout;
import org.apache.mxnet.javaapi.DType;
import org.apache.mxnet.javaapi.DataDesc;
import org.apache.mxnet.javaapi.NDArray;
import org.apache.mxnet.javaapi.Shape;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.List;

public class ObjectDetectorTest {

    private List<DataDesc> inputDesc;
    private BufferedImage inputImage;

    private List<List<ObjectDetectorOutput>> expectedResult;

    private ObjectDetector objectDetector;

    private int batchSize = 1;

    private int channels = 3;

    private int imageHeight = 512;

    private int imageWidth = 512;

    private String dataName = "data";

    private int topK = 5;

    private String predictedClassName = "lion"; // Random string

    private Shape getTestShape() {

        return new Shape(new int[] {batchSize, channels, imageHeight, imageWidth});
    }

    @Before
    public void setUp() {

        inputDesc = new ArrayList<>();
        inputDesc.add(new DataDesc(dataName, getTestShape(), DType.Float32(), Layout.NCHW()));
        inputImage = new BufferedImage(imageWidth, imageHeight, BufferedImage.TYPE_INT_RGB);
        objectDetector = Mockito.mock(ObjectDetector.class);
        expectedResult = new ArrayList<>();
        expectedResult.add(new ArrayList<ObjectDetectorOutput>());
        expectedResult.get(0).add(new ObjectDetectorOutput(predictedClassName, new float[]{}));
    }

    @Test
    public void testObjectDetectorWithInputImage() {

        Mockito.when(objectDetector.imageObjectDetect(inputImage, topK)).thenReturn(expectedResult);
        List<List<ObjectDetectorOutput>> actualResult = objectDetector.imageObjectDetect(inputImage, topK);
        Mockito.verify(objectDetector, Mockito.times(1)).imageObjectDetect(inputImage, topK);
        Assert.assertEquals(expectedResult, actualResult);
    }


    @Test
    public void testObjectDetectorWithBatchImage() {

        List<BufferedImage> batchImage = new ArrayList<>();
        batchImage.add(inputImage);
        Mockito.when(objectDetector.imageBatchObjectDetect(batchImage, topK)).thenReturn(expectedResult);
        List<List<ObjectDetectorOutput>> actualResult = objectDetector.imageBatchObjectDetect(batchImage, topK);
        Mockito.verify(objectDetector, Mockito.times(1)).imageBatchObjectDetect(batchImage, topK);
        Assert.assertEquals(expectedResult, actualResult);
    }

    @Test
    public void testObjectDetectorWithNDArrayInput() {

        NDArray inputArr = ObjectDetector.bufferedImageToPixels(inputImage, getTestShape());
        List<NDArray> inputL = new ArrayList<>();
        inputL.add(inputArr);
        Mockito.when(objectDetector.objectDetectWithNDArray(inputL, 5)).thenReturn(expectedResult);
        List<List<ObjectDetectorOutput>> actualResult = objectDetector.objectDetectWithNDArray(inputL, topK);
        Mockito.verify(objectDetector, Mockito.times(1)).objectDetectWithNDArray(inputL, topK);
        Assert.assertEquals(expectedResult, actualResult);
    }
}
